elki/src/main/java/de/lmu/ifi/dbs/elki/index/tree/spatial/ph/MinimalisticMemoryPHTree.java

```diff
@@ -0,0 +1,397 @@
+      distcalc.increment();
+    }
+  }
+
+  @Override
+  public KNNQuery<O> getKNNQuery(DistanceQuery<O> distanceQuery, Object... hints) {
+    DistanceFunction<? super O> df = distanceQuery.getDistanceFunction();
+    // TODO: if we know this works for other distance functions, add them, too!
+    if(df instanceof LPNormDistanceFunction) {
+      return new PHTreeKNNQuery(distanceQuery, (Norm<? super O>) df);
+    }
+    if(df instanceof SquaredEuclideanDistanceFunction) {
+      return new PHTreeKNNQuery(distanceQuery, (Norm<? super O>) df);
+    }
+    if(df instanceof SparseLPNormDistanceFunction) {
+      return new PHTreeKNNQuery(distanceQuery, (Norm<? super O>) df);
+    }
+    return null;
+  }
+
+  @Override
+  public RangeQuery<O> getRangeQuery(DistanceQuery<O> distanceQuery, Object... hints) {
+    DistanceFunction<? super O> df = distanceQuery.getDistanceFunction();
+    // TODO: if we know this works for other distance functions, add them, too!
+    if(df instanceof LPNormDistanceFunction) {
+      return new PHTreeRangeQuery(distanceQuery, (Norm<? super O>) df);
+    }
+    if(df instanceof SquaredEuclideanDistanceFunction) {
+      return new PHTreeRangeQuery(distanceQuery, (Norm<? super O>) df);
+    }
+    if(df instanceof SparseLPNormDistanceFunction) {
+      return new PHTreeRangeQuery(distanceQuery, (Norm<? super O>) df);
+    }
+    return null;
+  }
+
+  /**
+   * kNN query for the ph-tree.
+   * 
+   * @author Tilmann Zaeschke
+   */
+  public class PHTreeKNNQuery extends AbstractDistanceKNNQuery<O> {
+    /**
+     * Norm to use.
+     */
+    private Norm<? super O> norm;
+
+    /**
+     * Constructor.
+     * 
+     * @param distanceQuery Distance query
+     * @param norm Norm to use
+     */
+    public PHTreeKNNQuery(DistanceQuery<O> distanceQuery, Norm<? super O> norm) {
+      super(distanceQuery);
+      this.norm = norm;
+    }
+
+    @Override
+    public KNNList getKNNForObject(O obj, int k) {
+      final KNNHeap knns = DBIDUtil.newHeap(k);
+      
+      for (double[] v: tree.nearestNeighbour(0, oToDouble(obj, new double[dims]))) {
+        DBID id = tree.get(v);
+        O o2 = relation.get(id);
+        double distance = norm.distance(obj, o2);
+        knns.insert(distance, id);
+      }
+      
+      return knns.toKNNList();
+    }
+  }
+
+  /**
+   * Range query for the ph-tree.
+   * 
+   * In a range query, we return all entries within a given distance of a given point.
+   * 
+   * For the PH-Tree, we perform a query on the bounding rectangle and then filter out all
+   * points that still violate the distance function.
+   * 
+   * TODO
+   * Unfortunately, this does not scale well with high dimensions because in high dimensions,
+   * cubes grow exponentially faster than spheres.
+   * As a solution we should pass in the distance function into the PH-query mechanism where
+   * it can be used to skip sub-trees that would fit into the rectangle but not into the given
+   * distance. 
```

Comment:
This could be implemented with the PhTree, in fact I have a distance function that checks for dominance of vectors (SkyLine queries). It currently requires Java 8 because it is implemented as a lambda. Once ELKI is on Java 8, this should be easy to implement.

But as I wrote in the comment, high dimensional cubes grow exponentially faster than spheres in high dimensions. Skipping sub-trees in the PH-tree will help a bit, but I don;t think it can fully solve the dimensionality problem.